### ISSUE TITLE
Approve an allowance for an account or contract

### ISSUE THREAD

**Issue by SimiHunjan** [hashgraph/hedera-mirror-node-explorer#499]
### Problem

The Mirror Node Explorer currently does not support approving an allowance for tokens.

**User Stories**

As a developer or retail user, I want to approve a contract or account to spend hbars on behalf of my account. 

As a developer or retail user, I want to approve a contract or account to spend fungible on behalf of my account. 

As a developer or retail user, I want to approve a contract or account to spend non-fungible tokens on behalf of my account. 

As a developer or retail user, I want to see what contracts or accounts can spend hbars or fungible tokens on behalf of my account.

As a developer or retail user, I want to remove a contract or account allowance for hbars or fungible tokens.

### Solution

Add support for approving allowances.

**Flow**
- A user wants to allow a contract or account to spend tokens on their behalf
- The application provides a way for the user to enter the transaction details to approve the token allowance for a contract
    - The user specifies their account ID as the owner of the tokens
    - The user specifies the account ID or contract ID they want to grant the token approval for
    - The user specifies the ID of the token they want to grant the contract approval for
        - The token can be hbars, fungible token, an individual non-fungible token, or the entire class of non-fungible tokens
    - The method that needs to be called will depend on the token type the user specifies to grant the allowance for
        - If the user wants to grant an allowance for a fungible token, the corresponding method must be used
        - If the user wants to grant an allowance for a single NFT, the corresponding method must be used
        - If the user wants to grant an allowance for all NFTs in a given class, the corresponding method must be used
        - If the user wants to grant an allowance for hbars, the corresponding method must be used
    - The user specifies the number of tokens if it is hbar or a fungible token is specified 
- The transaction details will be sent to the supported wallet
- The user can sign the transaction using a supported wallet
- The user can view the contracts or accounts that were provided a spending allowance on behalf of the user account

**Comment by SimiHunjan** [comment]
The mirror node REST API does not yet return the NFTs an account has approved a spender for. 

https://github.com/hashgraph/hedera-mirror-node/issues/4348

**Comment by svienot** [comment]
Closing this  and opening #520 to track the missing piece of functionality which depends upon REST API enhancement.

**Comment by mshakeg** [comment]
@svienot could this be enabled for mainnet soon as many dApps have likely been affected by a change in [v0.36.0](https://github.com/hashgraph/hedera-services/releases/tag/v0.36.0) which requires accounts to approve token allowances.

https://github.com/hashgraph/hedera-services/issues/6253

**Comment by svienot** [comment]
Hi @mshakeg 
Approving  allowance is already supported by HashScan. You have to connect your wallet (on the Staking page) for the functionality to be visible:
<img width="1145" alt="Screenshot 2023-04-25 at 19 49 49" src="https://user-images.githubusercontent.com/16097111/234404782-574dfc24-3c63-404d-a870-58a94808f195.png">

**Comment by mshakeg** [comment]
Hi @svienot thanks.

I did notice an issue with allowances for NFTs not showing, for example on [account 0.0.1242](https://hashscan.io/testnet/account/0.0.1242).

I've approved an NFT allowance for:
1. ALL NFTs for [token 0.0.3974685](https://hashscan.io/testnet/token/0.0.3974685) in [this tx](https://hashscan.io/testnet/transaction/1682504065.622630003)
2. for NFT with `serial_id` 14 for the same token in [this tx](https://hashscan.io/testnet/transaction/1682504307.572309119)

Also, I'd like to recommend allowing for a user to select an infinite allowance for fungible or HBAR allowances. On Ethereum if an allowance is equal to `type(uint256).max` then that is understood to be infinite allowance. Not sure if this is even possible on Hedera but if it is I'd imagine `type(int64).max` to be understood as an infinite allowance.
